src/Workspaces/Core/Portable/Diagnostics/DiagnosticDataSerializer.cs

```diff
@@ -313,7 +313,7 @@ private static ImmutableArray<DiagnosticData> ReadDiagnosticDataArray(ObjectRead
 
             var documentId = document != null
                 ? document.Id
-                : project.Documents.FirstOrDefault(d => d.FilePath == originalFile)?.Id;
+                : project.State.DocumentStates.FirstOrDefault(d => d.Value.FilePath == originalFile).Key;
```

Comment:
Why isn't this just calling Solution.GetDocumentIdsByFilePath()? Then we're not using internal implementation details, and it's also not O(n).

Reply:
(we can still filter down to the IDs from the given project to avoid linked file issues)

Reply:
> Why isn't this just calling Solution.GetDocumentIdsByFilePath

I had no idea it existed 😄

Reply:
➡️ Now updated